dev-tools/reconfigurator-cli/src/main.rs

```diff
@@ -76,13 +81,66 @@ struct ReconfiguratorSim {
     /// External DNS zone name configured
     external_dns_zone_name: String,
 
+    /// RNG for collection IDs
+    collection_id_rng: TypedUuidRng<CollectionKind>,
+
     /// Policy overrides
     num_nexus: Option<u16>,
 
     log: slog::Logger,
 }
 
 impl ReconfiguratorSim {
+    fn new(log: slog::Logger) -> Self {
+        Self {
+            system: SystemDescription::new(),
+            collections: IndexMap::new(),
+            blueprints: IndexMap::new(),
+            internal_dns: BTreeMap::new(),
+            external_dns: BTreeMap::new(),
+            silo_names: vec!["default-silo".parse().unwrap()],
+            external_dns_zone_name: String::from("oxide.example"),
+            collection_id_rng: TypedUuidRng::from_entropy(),
+            num_nexus: None,
+            log,
+        }
+    }
+
+    /// Returns true if the REPL needs to be reset before loading an example.
+    ///
+    /// Loading example systems requires a (mostly) clean slate, which this
+    /// method checks for.
+    fn needs_reset_for_load_example(&self) -> bool {
+        // Use this pattern to ensure that if a new field is added to
+        // ReconfiguratorSim, it will fail to compile until it's added here.
+        let Self {
```

Comment:
This makes sense but I wonder if we can give more explicit instruction to somebody that's adding a new field?  How do they know which category it's in?

All of this has me wondering if we want to separate out these pieces into a separate object that we replace wholesale when needed...but I don't have a concrete suggestion at this point.  We can go with this and see how it goes.

Reply:
I started pulling this thread and it got out of hand a bit:

```
Cargo.lock                                                   |   16 +
Cargo.toml                                                   |    1 +
dev-tools/reconfigurator-cli/Cargo.toml                      |    3 +
dev-tools/reconfigurator-cli/src/dispatch.rs                 |   74 +++
dev-tools/reconfigurator-cli/src/lib.rs                      |   12 +
dev-tools/reconfigurator-cli/src/main.rs                     | 1310 +-------------------------------------------------
dev-tools/reconfigurator-cli/src/repl.rs                     | 1246 +++++++++++++++++++++++++++++++++++++++++++++++
dev-tools/reconfigurator-cli/src/rng.rs                      |  108 +++++
dev-tools/reconfigurator-cli/src/sim.rs                      |  346 +++++++++++++
dev-tools/reconfigurator-cli/tests/input/cmds-example.txt    |   26 +
dev-tools/reconfigurator-cli/tests/output/cmd-example-stderr |    0
dev-tools/reconfigurator-cli/tests/output/cmd-example-stdout |  526 ++++++++++++++++++++
dev-tools/reconfigurator-cli/tests/test_basic.rs             |   15 +
nexus/reconfigurator/planning/src/example.rs                 |    2 +-
nexus/reconfigurator/planning/src/system.rs                  |   12 +-
nexus/types/src/deployment/execution/dns.rs                  |    4 +-
16 files changed, 2390 insertions(+), 1311 deletions(-)
```

I'm going to tackle this in a separate PR.